waiter/src/waiter/scheduler/cook.clj

```diff
@@ -0,0 +1,529 @@
+  (defn- instance-healthy?
+    "Performs health check if an entry does not exist in the healthy-instance-cache."
+    [task-id health-check-url]
+    ;; TODO Move health check out of critical path, e.g. by storing a future in the cache
+    (let [health-result (utils/atom-cache-get-or-load
+                          healthy-instance-cache
+                          task-id
+                          (fn perform-instance-health-check []
+                            (try
+                              ;; no exception thrown means we got a 2XX response
+                              (log/info "performing health check for" task-id "at" health-check-url)
+                              (http-utils/http-request http-client health-check-url)
+                              (catch Exception e
+                                (log/error e "exception thrown while performing health check"
+                                           {:health-check-url health-check-url})
+                                false))))]
+      ;; Do not track unhealthy instances in the cache
+      (when-not health-result
+        (utils/atom-cache-evict healthy-instance-cache task-id))
+      health-result)))
+
+(defn- job->port
+  "Extracts the backend port from the job."
+  [job]
+  (if-let [backend-port (-> job :labels :backend-port)]
+    (Integer/parseInt backend-port)
+    (-> job :instances first :ports first)))
+
+(defn job-healthy?
+  "Returns true if the specified job has a healthy running instance."
+  [job]
+  (and (= "running" (-> job :status))
+       (= "running" (-> job :instances first :status))
+       (let [job-instance (-> job :instances first)
+             host (-> job-instance :hostname)
+             port (job->port job)
+             protocol (-> job :labels :backend-proto)
+             health-check-url (-> job :labels :health-check-url)
+             task-id (-> job-instance :task_id)]
+         (when (and host port protocol health-check-url)
+           (instance-healthy? task-id (str protocol "://" host ":" port health-check-url))))))
+
+;; Helper Methods
+
+(let [instance-expiry-adjustment-mins 5]
+  (defn create-job-description
+    "Create the Cook job description for a service."
+    [service-id service-description service-id->password-fn home-path-prefix instance-priority backend-port]
+    (let [{:strs [backend-proto cmd cmd-type cpus health-check-url instance-expiry-mins mem name ports
+                  run-as-user version]} service-description
+          job-uuid (str (UUID/randomUUID)) ;; TODO Use "less random" UUIDs for better Cook cache performance.
+          home-path (str home-path-prefix run-as-user)
+          environment (scheduler/environment service-id service-description service-id->password-fn home-path)
+          container-mode? (= "docker" cmd-type)
+          [_ image-namespace image-name image-label] (when container-mode?
+                                                       (re-matches #"(.*)/(.*):(.*)" version))
```

Comment:
I intentionally avoid performing this regex check earlier, e.g. during service description creation or token registration, to avoid overheads.